File: leetcode/src/main/java/com/giwankim/leetcode/SimplifyPath.java
package com.giwankim.leetcode;

import java.util.ArrayDeque;
import java.util.Deque;

public class SimplifyPath {
  /**
   * @implNote Time {@code O(n)}, space {@code O(n)}.
   */
  public String simplifyPath(String path) {
    String[] tokens = path.split("/");
    Deque<String> deque = new ArrayDeque<>();
    for (String token : tokens) {
      switch (token) {
        case "", "." -> {}
        case ".." -> deque.pollLast();
        default -> deque.addLast(token);
      }
    }
    return "/" + String.join("/", deque);
  }
}

File: leetcode-kt/src/main/kotlin/com/giwankim/leetcode/SimplifyPath.kt
package com.giwankim.leetcode

class SimplifyPath {
    fun simplifyPath(path: String): String {
        val stack = ArrayDeque<String>()
        val components = path.split("/")

        components.forEach {
            when (it) {
                "", "." -> {}
                ".." -> {
                    if (stack.isNotEmpty()) {
                        stack.removeLast()
                    }
                }
                else -> stack.addLast(it)
            }
        }

        val result =
            buildString {
                stack.forEach { append("/$it") }
            }
        return if (result.isNotEmpty()) result else "/"
    }
}
